service/frontend/adminHandler.go

```diff
@@ -654,50 +660,113 @@ func (adh *AdminHandler) DescribeCluster(_ context.Context, _ *adminservice.Desc
+		InitialFailoverVersion:   metadata.InitialFailoverVersion,
+		IsGlobalNamespaceEnabled: metadata.IsGlobalNamespaceEnabled,
 	}, nil
 }
 
-func (adh *AdminHandler) AddOrUpdateRemoteCluster(_ context.Context, _ *adminservice.AddOrUpdateRemoteClusterRequest) (_ *adminservice.AddOrUpdateRemoteClusterResponse, retError error) {
+func (adh *AdminHandler) AddOrUpdateRemoteCluster(
+	ctx context.Context,
+	request *adminservice.AddOrUpdateRemoteClusterRequest,
+) (_ *adminservice.AddOrUpdateRemoteClusterResponse, retError error) {
 	defer log.CapturePanic(adh.GetLogger(), &retError)
 
 	scope, sw := adh.startRequestProfile(metrics.AdminAddOrUpdateRemoteClusterScope)
 	defer sw.Stop()
 
-	/**
-	  1. Not allow update current cluster metadata (current cluster metadata can only be done from config)
-	  2. Call remote DescribeCluster
-	  3. Compare cluster_name, failover_version_increment, initial_failover_version, is_global_namespace_enabled
-	  4. persist the remote cluster metadata
-	*/
-	return nil, adh.error(serviceerror.NewUnimplemented("TODO"), scope)
+	adminClient, err := adh.Resource.GetClientFactory().NewAdminClientWithTimeout(
+		request.GetFrontendAddress(),
+		admin.DefaultTimeout,
+		admin.DefaultLargeTimeout,
+	)
+	if err != nil {
+		return nil, adh.error(err, scope)
+	}
+
+	// Fetch cluster metadata from remote cluster
+	resp, err := adminClient.DescribeCluster(ctx, &adminservice.DescribeClusterRequest{})
+	if err != nil {
+		return nil, adh.error(err, scope)
+	}
+
+	err = adh.validateRemoteClusterMetadata(resp)
+	if err != nil {
+		return nil, adh.error(err, scope)
+	}
+
+	clusterMetadataMrg := adh.GetClusterMetadataManager()
+	clusterData, err := clusterMetadataMrg.GetClusterMetadata(
+		&persistence.GetClusterMetadataRequest{ClusterName: resp.GetClusterName()},
+	)
+	if _, notFound := err.(*serviceerror.NotFound); err != nil && !notFound {
+		return nil, adh.error(err, scope)
+	}
+	var updateRequestVersion int64 = 0
+	if clusterData != nil {
+		updateRequestVersion = clusterData.Version
+	}
+
+	applied, err := clusterMetadataMrg.SaveClusterMetadata(&persistence.SaveClusterMetadataRequest{
+		ClusterMetadata: persistencespb.ClusterMetadata{
+			ClusterName:              resp.GetClusterName(),
+			HistoryShardCount:        resp.GetHistoryShardCount(),
+			ClusterId:                resp.GetClusterId(),
+			ClusterAddress:           request.GetFrontendAddress(),
+			FailoverVersionIncrement: resp.GetFailoverVersionIncrement(),
+			InitialFailoverVersion:   resp.GetInitialFailoverVersion(),
+			IsGlobalNamespaceEnabled: resp.GetIsGlobalNamespaceEnabled(),
+			IsConnectionEnabled:      request.GetEnableRemoteClusterConnection(),
+		},
+		Version: updateRequestVersion,
+	})
+	if err != nil {
+		return nil, adh.error(err, scope)
+	}
+	if !applied {
+		return nil, adh.error(serviceerror.NewInvalidArgument(
+			"Cannot add remote cluster due to update immutable fields"),
+			scope,
+		)
```

Comment:
would great to add which field violate the immutability

Reply:
This info would be good to pop up from the cluster metadata store. It just return a bool right now. We can return an error to have this information. However, I am not sure if this would be a API behavior change. I will take a look and see if I can do it in a separate PR.